src/file.cpp

```diff
@@ -1996,38 +2002,23 @@ typedef struct _FILE_ALLOCATED_RANGE_BUFFER {
 		bool ret = pAdjustTokenPrivileges(token, FALSE, &privs, 0, NULL, NULL)
 			&& GetLastError() == ERROR_SUCCESS;
 
-		CloseHandle(token);
-
 		return ret;
 	}
 
 	void set_file_valid_data(HANDLE f, boost::int64_t size)
 	{
 		typedef BOOL (WINAPI *SetFileValidData_t)(HANDLE, LONGLONG);
-		static SetFileValidData_t pSetFileValidData = NULL;
-		static bool failed_kernel32 = false;
+		static SetFileValidData_t const pSetFileValidData = (SetFileValidData_t)
+				GetProcAddress(GetModuleHandleA("kernel32"), "SetFileValidData");
 
-		if (pSetFileValidData == NULL && !failed_kernel32)
-		{
-			HMODULE k32 = LoadLibraryA("kernel32");
-			if (k32 == NULL)
-			{
-				failed_kernel32 = true;
-				return;
-			}
-			pSetFileValidData = (SetFileValidData_t)GetProcAddress(k32, "SetFileValidData");
-			if (pSetFileValidData == NULL)
-			{
-				failed_kernel32 = true;
-				return;
-			}
-		}
+		if (pSetFileValidData == NULL) return;
 
-		TORRENT_ASSERT(pSetFileValidData);
+		static bool const has_manage_volume_privs = get_manage_volume_privs();
```

Comment:
Postpone initialization of `has_manage_volume_privs`.

Reply:
technically this is a race condition in C++98, as static initializations aren't thread safe. But I would expect it to be a small problem in practice, especially given that mot popular compilers support C++11 by now.

Reply:
Also multiple calls to `get_manage_volume_privs()` from different threads aren't a problem, since they would return the same value.